File: Programmers/귤 고르기.py
# 귤을 크기 순서대로 오름차순 정렬을 한다.
# 연속적으로 크기가 같지 않은 경우에 new 배열에 담아준다.
# 반면 연속적으로 크기가 같으면 cnt를 카운트해주고 배열에 담지 않는다.
# cnt를 기준으로 내림차순 정렬을 하여 가장 많이 연속적으로 같은 크기의 귤을 선정한다.

def solution(k, tangerine):
    answer = 0
    tangerine.sort()
    new = []
    cnt = 1
    
    for i in range(len(tangerine)):
        if i == len(tangerine)-1:
            new.append([tangerine[i],cnt])
        elif tangerine[i] == tangerine[i+1]:
            cnt += 1
        else:
            new.append([tangerine[i],cnt])
            cnt = 1
    new.sort(key=lambda x:-x[1])
    
    total = 0
    for i in new:
        total += i[1]
        answer += 1
        
        if total >= k:
            return answer
        
    return answer

File: Programmers/귤 고르기.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int size;
    int count;
} Tangerine;

int compare_size(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

int compare_count(const void *a, const void *b) {
    return ((Tangerine *)b)->count - ((Tangerine *)a)->count;
}

int solution(int k, int *tangerine, int tangerineSize) {
    int answer = 0;
    qsort(tangerine, tangerineSize, sizeof(int), compare_size);
    
    Tangerine *new = (Tangerine *)malloc(tangerineSize * sizeof(Tangerine));
    int newSize = 0;
    int cnt = 1;
    
    for (int i = 0; i < tangerineSize; i++) {
        if (i == tangerineSize - 1) {
            new[newSize].size = tangerine[i];
            new[newSize].count = cnt;
            newSize++;
        } else if (tangerine[i] == tangerine[i + 1]) {
            cnt++;
        } else {
            new[newSize].size = tangerine[i];
            new[newSize].count = cnt;
            newSize++;
            cnt = 1;
        }
    }
    
    qsort(new, newSize, sizeof(Tangerine), compare_count);
    
    int total = 0;
    for (int i = 0; i < newSize; i++) {
        total += new[i].count;
        answer++;
        
        if (total >= k) {
            free(new);
            return answer;
        }
    }
    
    free(new);
    return answer;
}
